src/typecheck.rs

```diff
@@ -1,24 +1,98 @@
-//! The type checker
-
 use crate::syntax::*;
+use std::error::Error;
+use std::fmt;
 
-enum TypeError {
-    TypeMismatch,
+/// Errors that can occur during type checking
+#[derive(Debug)]
+pub enum TypeError {
+    /// Type mismatch between expected and actual types
+    TypeMismatch {
+        /// The type that was expected
+        expected: Type,
+        /// The type that was actually found
+        actual: Type,
+    },
+    /// Variable not found in the environment
+    UndefinedVariable(Id),
+    /// Function not found
+    UndefinedFunction(Id),
+    /// Symbol not in alphabet
+    SymbolNotInAlphabet(char),
 }
 
-// TODO how to sturcture the type checker
-// TODO Casting with more concrete examples
-
-
-fn typeck_expr(expr: &Expr) -> Result<(), TypeError> {
-    todo!()
+// Implement Display for TypeError
+impl fmt::Display for TypeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            TypeError::TypeMismatch { expected, actual } => {
+                write!(
+                    f,
+                    "Type mismatch: expected {:?}, found {:?}",
+                    expected, actual
+                )
+            }
+            TypeError::UndefinedVariable(id) => {
+                write!(f, "Undefined variable: '{}'", id)
+            }
+            TypeError::UndefinedFunction(id) => {
+                write!(f, "Undefined function: '{}'", id)
+            }
+            TypeError::SymbolNotInAlphabet(c) => {
+                write!(f, "Symbol '{}' is not in the alphabet", c)
+            }
+        }
+    }
 }
 
-fn typeck_stmt(stmt: &Stmt) -> Result<(), TypeError> {
-    todo!()
-}
+// Implement Error for TypeError
+impl Error for TypeError {}
```

Comment:
You can derive `Error` automatically using https://docs.rs/thiserror/latest/thiserror/